### PR TITLE
feat: Improve add fn key estimation

### PR DESCRIPTION
**Pull request by jakmeier**
Test short and long method names in the gas cost estimation for adding a
function call access. Also subtract the receipt creation cost.

### CODE REVIEW

File: runtime/runtime-params-estimator/src/lib.rs [hunk 1]
```diff
@@ -399,28 +399,53 @@ fn action_add_function_access_key_base(ctx: &mut EstimatorContext) -> GasCost {
 }
 
 fn action_add_function_access_key_per_byte(ctx: &mut EstimatorContext) -> GasCost {
-    let total_cost = {
-        let many_methods: Vec<_> = (0..1000).map(|i| format!("a123456{:03}", i)).collect();
+    let base_cost = action_add_function_access_key_base(ctx) + action_sir_receipt_creation(ctx);
+
+    // Set up estimation with varying method length and total bytes.
+    let mut estimate = |method_len: usize, total_len: usize| {
```
[the remark this answers is not in the document]
Comment (matklad) [reply runtime/runtime-params-estimator/src/lib.rs:405]: Otoh, I'd just write this as a loop, which has exactly the same problem :) 

```
let cases = [(max_method_len, max_bytes), ]
let mut estimations = Vec::with_capacity(cases.len());
for (m_len, n_methods) {
   ....
}
```

File: runtime/runtime-params-estimator/src/lib.rs [hunk 2]
```diff
@@ -399,28 +399,53 @@ fn action_add_function_access_key_base(ctx: &mut EstimatorContext) -> GasCost {
...
+        // Nothing prevents a key to list the same method many times. Performance should not be affected.
+        let method_name = "x".repeat(method_len);
```
Comment (matklad) [inline runtime/runtime-params-estimator/src/lib.rs:407]: I'd maybe still made names different, just in case something does change in the future impl.

**Review by matklad — APPROVED** [review]

File: runtime/runtime-params-estimator/src/lib.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (jakmeier) [reply runtime/runtime-params-estimator/src/lib.rs:405]: I had the same thought. Almost made it a loop, almost put it in a separate function. So I definitely see where you are coming form.

But IMO in this specific case it is just more readable this way. (For programmers that are not spooked by closures, that is.)

File: runtime/runtime-params-estimator/src/lib.rs (hunk 2 above)
Comment (jakmeier) [reply runtime/runtime-params-estimator/src/lib.rs:407]: Well, if we make names different, names need to be longer than one character.
But an attacker is not restricted to that today, so it is totally feasible to submit 1000x "x". And we estimate that case as slightly more gas heavy than fewer methods with longer names.

**Pull request merged**